File: src/lootscraper/html.py
# mypy: ignore-errors
import html
import logging
from collections.abc import Sequence
from datetime import datetime, timezone
from pathlib import Path

from jinja2 import Template

from lootscraper.database import Game, Offer

from .common import (
    TIMESTAMP_SHORT,
    OfferDuration,
    OfferType,
    Source,
)

logger = logging.getLogger(__name__)

TEMPLATE_STR = """
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>{{ feed.title }}</title>
    <link href="https://cdn.jsdelivr.net/npm/tailwindcss@2.2.19/dist/tailwind.min.css" rel="stylesheet">
</head>
<body class="bg-gray-100">
    <div class="container mx-auto p-8">
        <h1 class="text-4xl font-bold mb-4">{{ feed.title }}</h1>

        {% for entry in entries|sort(attribute='valid_from', reverse=true) %}

        <!-- Header & Validity -->
        {% if entry.is_expired %}
        <h2 class="flex justify-between items-center text-2xl mb-2 line-through">
        {% else %}
        <h2 class="flex justify-between items-center text-2xl mb-2">
        {% endif %}

        <span class="font-bold">{{ entry.title }}</span>
        {% if entry.valid_from and entry.valid_to %}
        <span class="text-sm">Valid from {{ entry.valid_from }} to {{ entry.valid_to }}</span>
        {% endif %}
        </h2>

        <div class="flex bg-white rounded-lg shadow-md p-6 mb-8">
            <!-- Image -->
            <div>
                {% if entry.img_url %}
                <img src="{{ entry.img_url }}" alt="{{ entry.title }}" class="w-96">
                {% endif %}
            </div>

            <!-- Content -->
            <div class="flex-1 ml-4">

                <!-- Game Info -->
                <div class="bg-gray-200 p-4 rounded-lg text-sm">
                    <div class="flex justify-between mb-2">
                    <h3 class="font-bold underline">
                        {{ entry.game_name }}
                    </h3>
                    {% if entry.release_date %}
                    <span>
                        <strong>Release Date:</strong> {{ entry.release_date }}
                    </span>
                    {% endif %}
                    </div>


                    <!-- Ratings -->
                    <div class="flex flex-wrap mb-2">
                        {% if entry.steam_percent %}
                        <span class="rounded-full px-3 py-1 text-sm font-semibold m-1
                        {% if entry.steam_percent > 90 %}bg-green-700
                        {% elif entry.steam_percent > 80 %}bg-green-500
                        {% elif entry.steam_percent > 60 %}bg-yellow-500
                        {% else %}bg-red-600{% endif %}">
                            Steam: {{ entry.steam_percent }}% / {{ entry.steam_score }} ({{ entry.steam_recommendations }} recommendations)
                        </span>
                        {% endif %}
                        {% if entry.igdb_meta_score %}
                        <span class="rounded-full px-3 py-1 text-sm font-semibold m-1
                        {% if entry.igdb_meta_score > 90 %}bg-green-700
                        {% elif entry.igdb_meta_score > 80 %}bg-green-500
                        {% elif entry.igdb_meta_score > 60 %}bg-yellow-500
                        {% else %}bg-red-600{% endif %}">
                            IGDB Meta: {{ entry.igdb_meta_score }}% ({{ entry.igdb_meta_ratings }} sources)
                        </span>
                        {% endif %}
                        {% if entry.metacritic_score %}
                        <span class="rounded-full px-3 py-1 text-sm font-semibold m-1
                        {% if entry.metacritic_score > 90 %}bg-green-700
                        {% elif entry.metacritic_score > 80 %}bg-green-500
                        {% elif entry.metacritic_score > 60 %}bg-yellow-500
                        {% else %}bg-red-600{% endif %}">
                            Metacritic: {{ entry.metacritic_score }}%
                        </span>
                        {% endif %}
                        {% if entry.igdb_user_score %}
                        <span class="rounded-full px-3 py-1 text-sm font-semibold m-1
                        {% if entry.igdb_user_score > 90 %}bg-green-700
                        {% elif entry.igdb_user_score > 80 %}bg-green-500
                        {% elif entry.igdb_user_score > 60 %}bg-yellow-500
                        {% else %}bg-red-600{% endif %}">
                            IGDB User: {{ entry.igdb_user_score }}% ({{ entry.igdb_user_ratings }} sources)
                        </span>
                        {% endif %}
                    </div>

                    <!-- Genres -->
                    <div class="flex flex-wrap mb-2">
                        {% if entry.genres %}
                        <div class="flex flex-wrap">
                            {% for genre in entry.genres.split(', ') %}
                            <span class="text-white bg-gray-500 rounded-full px-3 py-1 text-sm font-semibold m-1">#{{ genre }}</span>
                            {% endfor %}
                        </div>
                        {% else %}
                        <span class="text-white bg-gray-500 rounded-full px-3 py-1 text-sm font-semibold m-1">#Unknown</span>
                        {% endif %}
                    </div>

                    <!-- Other Info -->
                    <div>
                        {{ entry.description }}
                    </div>


                </div>

                <!-- Claim Button -->
                <div class="flex justify-between items-center mt-4">
                    <!-- Claim Button -->
                    {% if entry.url %}
                    <a href="{{ entry.url }}" class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded">
                        Claim on {{ entry.source }}
                    </a>
                    {% endif %}

                    <!-- Recommended Price -->
                    {% if entry.recommended_price %}
                    <span class="bg-gray-500 text-white font-bold py-2 px-4 rounded line-through">
                        {{ entry.recommended_price }} EUR
                    </span>
                    {% endif %}
                </div>

            </div>
        </div>
        {% endfor %}
    </div>
</body>
</html>
"""  # noqa: E501


def generate_html(
    offers: Sequence[Offer],
    file: Path,
    author_name: str,
    author_mail: str,
    author_web: str,
    feed_id_prefix: str,
    source: Source | None = None,
    type_: OfferType | None = None,
    duration: OfferDuration | None = None,
) -> None:
    """Generate a html view with the given offers."""
    if len(offers) == 0:
        return

    latest_date: datetime | None = None

    entries = []

    for offer in offers:
        # Skip entries without any dates, they are probably not valid
        if not offer.seen_first and not offer.seen_last:
            continue

        # Skip future entries and entries that are no longer seen before they
        # ever were valid
        if offer.valid_from and offer.valid_from > offer.seen_last:
            continue

        # Determine the date to use for updated. Preferred is valid_from,
        # but seen_first is used if valid_from is not known.
        if (
            offer.valid_from
            and offer.seen_first
            and offer.valid_from > offer.seen_first
        ):
            updated: datetime | None = offer.valid_from
        else:
            updated = offer.seen_first

        # Remember the newest entry date for the whole feed date
        if not latest_date or (updated and updated > latest_date):
            latest_date = updated

        # Save the needed info

        entry = {}
        entry["id"] = f"{feed_id_prefix}{int(offer.id)}"

        additional_info = offer.type.value
        if offer.duration != OfferDuration.CLAIMABLE:
            additional_info += f", {offer.duration.value}"

        entry["title"] = offer.title

        # - Content
        game: Game | None = offer.game

        if offer.img_url:
            entry["img_url"] = html.escape(offer.img_url)
        elif game and game.steam_info and game.steam_info.image_url:
            entry["img_url"] = html.escape(game.steam_info.image_url)

        valid_from = (
            offer.valid_from.strftime(TIMESTAMP_SHORT)
            if offer.valid_from
            else offer.seen_first.strftime(TIMESTAMP_SHORT)
        )
        entry["valid_from"] = valid_from
        if offer.valid_to:
            entry["valid_to"] = offer.valid_to.strftime(TIMESTAMP_SHORT)
        if offer.url:
            entry["source"] = html.escape(offer.source.value)
            entry["url"] = html.escape(offer.url)

        if offer.valid_to and offer.valid_to < datetime.now(tz=timezone.utc):
            entry["is_expired"] = True
        else:
            entry["is_expired"] = False

        if game:
            entry["has_game"] = True
            if game.igdb_info and game.igdb_info.name:
                entry["game_name"] = html.escape(game.igdb_info.name)
            elif game.steam_info and game.steam_info.name:
                entry["game_name"] = html.escape(game.steam_info.name)

            if game.steam_info and game.steam_info.metacritic_score:
                entry["metacritic_score"] = game.steam_info.metacritic_score
                if game.steam_info.metacritic_url:
                    entry["metacritic_url"] = html.escape(
                        game.steam_info.metacritic_url,
                    )
            if (
                game.steam_info
                and game.steam_info.percent
                and game.steam_info.score
                and game.steam_info.recommendations
            ):
                entry["steam_percent"] = game.steam_info.percent
                entry["steam_score"] = game.steam_info.score
                entry["steam_recommendations"] = game.steam_info.recommendations
                entry["steam_url"] = html.escape(game.steam_info.url)
            if (
                game.igdb_info
                and game.igdb_info.meta_ratings
                and game.igdb_info.meta_score
            ):
                entry["igdb_meta_score"] = game.igdb_info.meta_score
                entry["igdb_meta_ratings"] = game.igdb_info.meta_ratings
                entry["igdb_url"] = html.escape(game.igdb_info.url)
            if (
                game.igdb_info
                and game.igdb_info.user_ratings
                and game.igdb_info.user_score
            ):
                entry["igdb_user_score"] = game.igdb_info.user_score
                entry["igdb_user_ratings"] = game.igdb_info.user_ratings
            if game.igdb_info and game.igdb_info.release_date:
                entry["release_date"] = game.igdb_info.release_date.strftime(
                    TIMESTAMP_SHORT,
                )
            elif game.steam_info and game.steam_info.release_date:
                entry["release_date"] = game.steam_info.release_date.strftime(
                    TIMESTAMP_SHORT,
                )
            if game.steam_info and game.steam_info.recommended_price_eur:
                entry["recommended_price"] = game.steam_info.recommended_price_eur
            if game.igdb_info and game.igdb_info.short_description:
                entry["description"] = html.escape(
                    game.igdb_info.short_description,
                )
            elif game.steam_info and game.steam_info.short_description:
                entry["description"] = html.escape(
                    game.steam_info.short_description,
                )
            if game.steam_info and game.steam_info.genres:
                entry["genres"] = html.escape(game.steam_info.genres)

        # Add to array
        entries.append(entry)

    feed = {
        "author_name": author_name,
        "author_email": author_mail,
        "author_uri": author_web,
        "title": get_title(source, type_, duration),
        "updated": latest_date,
    }

    template = Template(TEMPLATE_STR)
    rendered_html = template.render(entries=entries, feed=feed)

    with file.open("w", encoding="utf-8") as f:
        f.write(rendered_html)


def get_title(
    source: Source | None,
    type_: OfferType | None,
    duration: OfferDuration | None,
) -> str:
    if source is None and type_ is None and duration is None:
        return "Free Games and Loot"

    title = "Free"

    if source is not None:
        title += " " + source.value

    if type_ == OfferType.GAME:
        title += " Games"
    elif type_ == OfferType.LOOT:
        title += " Loot"

    if duration in (OfferDuration.TEMPORARY, OfferDuration.ALWAYS):
        title += f" ({duration.value})"

    return title

File: src/services/generators/html.ts
import { writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import Handlebars from "handlebars";
import { DateTime } from "luxon";
import { getGameWithInfo } from "@/services/database/gameRepository";
import type { FeedCombination } from "@/services/scraper/utils";
import { translationService } from "@/services/translation";
import type { Config } from "@/types/config";
import type { Offer } from "@/types/database";
import { getDataPath } from "@/utils/path";
import {
  cleanHtml,
  generateFeedTitle,
  generateFilename,
} from "@/utils/stringTools";

// The latest static URL for the Tailwind CSS stylesheet.
const CSS_URL =
  "https://cdn.jsdelivr.net/npm/tailwindcss@2.2.19/dist/tailwind.min.css";
const TEMPLATE = `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>{{feed.title}}</title>
    <link href="${CSS_URL}" rel="stylesheet">
</head>
<body class="bg-gray-100">
    <div class="container mx-auto p-8">
        <h1 class="text-4xl font-bold mb-4">{{feed.title}}</h1>

        {{#each entries}}
        <!-- Header & Validity -->
        <h2 class="flex justify-between items-center text-2xl mb-2 {{#if is_expired}}line-through{{/if}}">
            <span class="font-bold">{{{title}}}</span>
            <span class="text-sm">Valid from {{{valid_from}}}{{#if valid_to}} to {{{valid_to}}}{{/if}}</span>
        </h2>

        <div class="flex bg-white rounded-lg shadow-md p-6 mb-8">
            {{#if img_url}}
            <!-- Image -->
            <div>
                <img src="{{{img_url}}}" alt="{{{title}}}" class="w-96">
            </div>
            {{/if}}

            <!-- Content -->
            <div class="flex-1 ml-4">
                {{#if has_game}}
                <!-- Game Info -->
                <div class="bg-gray-200 p-4 rounded-lg text-sm">
                    <div class="flex justify-between mb-2">
                        <h3 class="font-bold underline">{{{game_name}}}</h3>
                        {{#if release_date}}
                        <span><strong>Release Date:</strong> {{{release_date}}}</span>
                        {{/if}}
                    </div>

                    <!-- Ratings -->
                    <div class="flex flex-wrap mb-2">
                        {{#if steam_percent}}
                        <span class="rounded-full px-3 py-1 text-sm font-semibold m-1
                            {{#if (gt steam_percent 90)}}bg-green-700
                            {{else if (gt steam_percent 80)}}bg-green-500
                            {{else if (gt steam_percent 60)}}bg-yellow-500
                            {{else}}bg-red-600{{/if}}">
                            <a href="{{{steam_url}}}">Steam: {{steam_percent}}% / {{steam_score}} ({{steam_recommendations}} recommendations)</a>
                        </span>
                        {{/if}}

                        {{#if igdb_meta_score}}
                        <span class="rounded-full px-3 py-1 text-sm font-semibold m-1
                            {{#if (gt igdb_meta_score 90)}}bg-green-700
                            {{else if (gt igdb_meta_score 80)}}bg-green-500
                            {{else if (gt igdb_meta_score 60)}}bg-yellow-500
                            {{else}}bg-red-600{{/if}}">
                            <a href="{{{igdb_url}}}">IGDB Meta: {{igdb_meta_score}}% ({{igdb_meta_ratings}} sources)</a>
                        </span>
                        {{/if}}

                        {{#if metacritic_score}}
                        <span class="rounded-full px-3 py-1 text-sm font-semibold m-1
                            {{#if (gt metacritic_score 90)}}bg-green-700
                            {{else if (gt metacritic_score 80)}}bg-green-500
                            {{else if (gt metacritic_score 60)}}bg-yellow-500
                            {{else}}bg-red-600{{/if}}">
                            {{#if metacritic_url}}<a href="{{{metacritic_url}}}">{{/if}}
                            Metacritic: {{metacritic_score}}%
                            {{#if metacritic_url}}</a>{{/if}}
                        </span>
                        {{/if}}

                        {{#if igdb_user_score}}
                        <span class="rounded-full px-3 py-1 text-sm font-semibold m-1
                            {{#if (gt igdb_user_score 90)}}bg-green-700
                            {{else if (gt igdb_user_score 80)}}bg-green-500
                            {{else if (gt igdb_user_score 60)}}bg-yellow-500
                            {{else}}bg-red-600{{/if}}">
                            <a href="{{{igdb_url}}}">IGDB User: {{igdb_user_score}}% ({{igdb_user_ratings}} sources)</a>
                        </span>
                        {{/if}}
                    </div>

                    {{#if genres}}
                    <!-- Genres -->
                    <div class="flex flex-wrap mb-2">
                        <div class="flex flex-wrap">
                            {{#each (split genres ", ")}}
                            <span class="text-white bg-gray-500 rounded-full px-3 py-1 text-sm font-semibold m-1">#{{{this}}}</span>
                            {{/each}}
                        </div>
                    </div>
                    {{/if}}

                    {{#if description}}
                    <!-- Description -->
                    <div>{{description}}</div>
                    {{/if}}
                </div>
                {{/if}}

                <div class="flex justify-between items-center mt-4">
                    {{#if url}}
                    <!-- Claim Button -->
                    <a href="{{{url}}}" class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded">
                        Claim on {{{source}}}
                    </a>
                    {{/if}}

                    {{#if recommended_price}}
                    <!-- Price -->
                    <span class="bg-gray-500 text-white font-bold py-2 px-4 rounded line-through">
                        {{recommended_price}} EUR
                    </span>
                    {{/if}}
                </div>
            </div>
        </div>
        {{/each}}
    </div>
</body>
</html>`;

interface HtmlGeneratorOptions {
  combination?: FeedCombination;
  withHistory?: boolean;
}

export class HtmlGenerator {
  private readonly template: HandlebarsTemplateDelegate;

  constructor(
    private readonly config: Config,
    private readonly options: HtmlGeneratorOptions = {},
  ) {
    this.registerHelpers();
    this.template = Handlebars.compile(TEMPLATE);
  }

  private registerHelpers(): void {
    Handlebars.registerHelper("gt", (a: number, b: number) => a > b);
    Handlebars.registerHelper("split", (str: string, separator: string) =>
      str.split(separator),
    );
  }

  public async generateHtml(offers: Offer[]): Promise<void> {
    const entries = [];
    for (const offer of offers) {
      const gameInfo = offer.game_id
        ? await getGameWithInfo(offer.game_id)
        : null;

      entries.push({
        id: `${this.config.feed.idPrefix}${offer.id.toFixed()}`,
        title: offer.title,
        img_url: offer.img_url ?? gameInfo?.steamInfo?.image_url,
        valid_from: offer.valid_from
          ? DateTime.fromISO(offer.valid_from).toFormat("yyyy-MM-dd")
          : DateTime.fromISO(offer.seen_first).toFormat("yyyy-MM-dd"),
        valid_to: offer.valid_to
          ? DateTime.fromISO(offer.valid_to).toFormat("yyyy-MM-dd")
          : undefined,
        source: translationService.getSourceDisplay(offer.source),
        url: offer.url,
        is_expired:
          offer.valid_to && DateTime.fromISO(offer.valid_to) < DateTime.now(),
        has_game: !!gameInfo,
        game_name: gameInfo?.igdbInfo?.name ?? gameInfo?.steamInfo?.name,
        metacritic_score: gameInfo?.steamInfo?.metacritic_score,
        metacritic_url: gameInfo?.steamInfo?.metacritic_url,
        steam_percent: gameInfo?.steamInfo?.percent,
        steam_score: gameInfo?.steamInfo?.score,
        steam_recommendations: gameInfo?.steamInfo?.recommendations,
        steam_url: gameInfo?.steamInfo?.url,
        igdb_meta_score: gameInfo?.igdbInfo?.meta_score,
        igdb_meta_ratings: gameInfo?.igdbInfo?.meta_ratings,
        igdb_url: gameInfo?.igdbInfo?.url,
        igdb_user_score: gameInfo?.igdbInfo?.user_score,
        igdb_user_ratings: gameInfo?.igdbInfo?.user_ratings,
        release_date: gameInfo?.igdbInfo?.release_date
          ? DateTime.fromISO(gameInfo.igdbInfo.release_date).toFormat(
              "yyyy-MM-dd",
            )
          : gameInfo?.steamInfo?.release_date
            ? DateTime.fromISO(gameInfo.steamInfo.release_date).toFormat(
                "yyyy-MM-dd",
              )
            : undefined,
        recommended_price: gameInfo?.steamInfo?.recommended_price_eur,
        description:
          gameInfo?.igdbInfo?.short_description ??
          gameInfo?.steamInfo?.short_description,
        genres: gameInfo?.steamInfo?.genres,
      });
    }

    const html = this.template({
      feed: {
        title: this.getFeedTitle(),
        author_name: this.config.feed.authorName,
        author_email: this.config.feed.authorEmail,
        author_web: this.config.feed.authorWeb,
      },

      // Sort time entries in descending order by valid_from date,
      // then valid_to date (if present), and finally in ascending order by title.

      entries: entries.sort((a, b) => {
        // First compare by valid_from
        const validFromComparison = b.valid_from.localeCompare(a.valid_from);

        if (validFromComparison !== 0) {
          return validFromComparison;
        }

        // If valid_from is the same, compare by valid_to (if it exists)
        if (a.valid_to && b.valid_to) {
          const validToComparison = b.valid_to.localeCompare(a.valid_to);

          if (validToComparison !== 0) {
            return validToComparison;
          }
        }

        // If valid_to is also the same, compare by title
        return a.title.localeCompare(b.title);
      }),
    });

    const cleanedHtml = cleanHtml(html);

    const outputPath = resolve(getDataPath(), this.getFilename());
    await writeFile(outputPath, cleanedHtml);
  }

  private getFilename(): string {
    return generateFilename({
      prefix: this.config.common.feedFilePrefix,
      extension: "html",
      ...(this.options.combination && {
        combination: this.options.combination,
      }),
      ...(this.options.withHistory && {
        withHistory: this.options.withHistory,
      }),
    });
  }

  private getFeedTitle(): string {
    return generateFeedTitle(this.options.combination);
  }
}
